diff -ruN linux-4.6/kernel/fork.c uksm-0.1.2.4-beta-linux-4.6/kernel/fork.c
--- linux-4.6/kernel/fork.c	2016-05-16 06:43:13.000000000 +0800
+++ uksm-0.1.2.4-beta-linux-4.6/kernel/fork.c	2016-05-23 20:09:16.065570208 +0800
@@ -455,7 +455,7 @@
 				goto fail_nomem;
 			charge = len;
 		}
-		tmp = kmem_cache_alloc(vm_area_cachep, GFP_KERNEL);
+		tmp = kmem_cache_zalloc(vm_area_cachep, GFP_KERNEL);
 		if (!tmp)
 			goto fail_nomem;
 		*tmp = *mpnt;
@@ -508,7 +508,7 @@
 		__vma_link_rb(mm, tmp, rb_link, rb_parent);
 		rb_link = &tmp->vm_rb.rb_right;
 		rb_parent = &tmp->vm_rb;
-
+		uksm_vma_add_new(tmp);
 		mm->map_count++;
 		retval = copy_page_range(mm, oldmm, mpnt);
 
